control-flow/index.md

```diff
@@ -1,6 +1,368 @@
+
+```
+runOutPromise() // returns a response
+  .then(response => {
+    return foo; // this could optionally be another promise
+  })
+  .then(foo => {
+    // they can be chained
+  })
+  .catch(error => {
+    // catch "reject" output here
+  });
+```
+
+We can rewrite our runner example using promises like so:
+
+```
+function randomDelayedResponse(text, callback) {
+  return new Promise((resolve, reject) => {
+    const timeOut = Math.floor(Math.random() * 10) + 1;
+    setTimeout(() => {
+      resolve(text); // Replacing the callback with "resolve"
+    }, timeOut * 10);
+  });
+}
+
+randomDelayedResponse(1)
+  .then(res => {
+    console.log(res);
+    return randomDelayedResponse(2);
+  })
+  .then(res => {
+    console.log(res);
+    return randomDelayedResponse(3);
+  })
+  .then(res => {
+    console.log(res);
+    return randomDelayedResponse(4);
+  })
+  .then(res => {
+    console.log(res);
+  })
+  .catch(error => {
+    // handle error
+  })
+// outputs "1 2 3 4"
+```
+
+Running the above code should output the correct order, albeit with random delays between each number.
+
+We aren't handling errors in this example, but if our function was to encounter an error we would `reject('some error')` and this would then be picked up by the `catch`.
+
+You can learn more about `promises` here (LINK)
+
+Promises remove a lot of nesting and give us easier to read code. However there is a third way to control the execution order of our code.
+
+## Async / Await
+
+The third approach is built on top of the existing _promises_ approach but makes it easier to reason with. With `async` and `await` we can write code that feels a lot more like the simple top-down code, by telling it to wait when we need it to. Let's rewrite our _who will win?_ example from above.
+
+```
+function randomDelayedResponse(text, callback) {
+  return new Promise((resolve, reject) => {
+    const timeOut = Math.floor(Math.random() * 10) + 1;
+    setTimeout(() => {
+      resolve(text);
+    }, timeOut * 10);
+  });
+}
+
+async function runTheRace() { // put `async` before the function call
+
+  console.log(1) // Not async
+
+  const result2 = await randomDelayedResponse(2); // async
+  console.log(result2);
+
+  console.log(3);
+
+  const result4 = await randomDelayedResponse(4);
+  console.log(result4)
+}
+
+runTheRace();
+```
+
+In the above we have replaced the series of `.then()` calls with an `async` function. It's still making use of a _promise_ but when we want to wait for the promise we simply place `await` before the asynchronous function. This way, the code execution order is preserved.
+
+You may be wondering how we handle errors here. One approach is to use the `try/catch` method. We'll be covering that and more in the [error handling section]({{ "/error-handling/" | url }}).
+
```

Comment:
Worth adding a comment saying async/awaits is commonly preferred, as it helps promote clean code that tends to be easier to follow than the callback and promise patterns?